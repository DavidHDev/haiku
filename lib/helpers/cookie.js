export const parseToDataType = (value, isItRetry = false) => {
  try {
    return value === 'undefined' ? undefined : JSON.parse(value ?? '');
  } catch (e) {
    if (!isItRetry) return parseToDataType(`"${value?.replaceAll?.('"', '')}"`, true);
    return undefined;
  }
};

export const parseToCookieType = (value) => {
  if(typeof value === 'string') return value;

  return JSON.stringify(value);
}

export const getCookie = (name) => {
  const value = `; ${document.cookie}`;

  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType(cookie.split(';')[0]) : cookie;
};

export const getCookies = (cookies = []) => {
  if (cookies.length) return cookies.reduce((result, cookie) => ({ ...result, [cookie]: getCookie(cookie) }), {});

  return Object.fromEntries(document.cookie.split('; ').map((c) => {
    const [key, value] = c.split('=');
    return [key, parseToDataType(value)]
  }));
};

export const setCookie = (name, value, expireDays) => {
  const date = new Date();

  date.setTime(date.getTime() + expireDays * 24 * 60 * 60 * 1000);
  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name) => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
