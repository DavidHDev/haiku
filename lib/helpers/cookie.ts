export const parseToDataType = <T>(
  value: string | undefined,
  isItRetry = false,
): T | undefined => {
  try {
    return value === 'undefined' || value === undefined
      ? undefined
      : (JSON.parse(value) as T);
  } catch (e) {
    if (!isItRetry) {
      return parseToDataType<T>(`"${value?.replaceAll?.('"', '')}"`, true);
    }
    return undefined;
  }
};

export const parseToCookieType = <T>(value: T): string => {
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
};

export const getCookie = <T>(name: string): T | undefined => {
  const value = `; ${document.cookie}`;
  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType<T>(cookie.split(';')[0]) : undefined;
};

export const getCookies = <T>(
  cookies: string[] = [],
): Record<string, T | undefined> => {
  if (cookies.length) {
    return cookies.reduce(
      (result, cookie) => ({ ...result, [cookie]: getCookie<T>(cookie) }),
      {} as Record<string, T | undefined>,
    );
  }

  return Object.fromEntries(
    document.cookie.split('; ').map((c) => {
      const [key, value] = c.split('=');

      return [key, parseToDataType<T>(value)];
    }),
  ) as Record<string, T | undefined>;
};

export const setCookie = <T>(
  name: string,
  value: T,
  expireDays: number,
): void => {
  const date = new Date();
  const millisecondsInADay = 24 * 60 * 60 * 1000;
  const expireDate = date.getTime() + expireDays * millisecondsInADay;

  date.setTime(expireDate);

  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name: string): void => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
