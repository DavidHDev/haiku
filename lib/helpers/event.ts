export const on = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.addEventListener)
    target.addEventListener(type, callback, options);
};

export const off = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.removeEventListener)
    target.removeEventListener(type, callback, options);
};
