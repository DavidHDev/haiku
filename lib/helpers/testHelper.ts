import { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { expect } from "vitest";

export const takeSnapshot = (ui: ReactNode) => {
  const { asFragment } = rtlRender(ui);
  expect(asFragment()).toMatchSnapshot();
};
