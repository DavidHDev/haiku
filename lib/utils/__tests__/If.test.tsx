import { If } from '../If';
import { describe, expect, it } from 'vitest';
import { takeSnapshot } from 'lib/helpers/testHelper';
import { render, screen } from '@testing-library/react';

describe('If', () => {
  it('renders the If component', () => {
    takeSnapshot(<If isTrue>Hello</If>);
    expect(screen.getByText('Hello')).toBeInTheDocument();
  });

  it('should not render anything if isTrue is false', () => {
    render(<If isTrue={false}>Hello</If>);
    expect(screen.queryByText('Hello')).toBeNull();
    expect(screen.queryByText('Hello')).not.toBeInTheDocument();
  });
});
