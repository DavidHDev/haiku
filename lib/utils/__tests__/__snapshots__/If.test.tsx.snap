// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`If > renders the If component 1`] = `
<DocumentFragment>
  Hello
</DocumentFragment>
`;
